wincrypto/src/dtls.rs

```diff
@@ -0,0 +1,616 @@
+                self.state = EstablishmentState::Failed;
+                Err(WinCryptoError(format!("DTLS handshake failure: {:?}", e)).into())
+            }
+        };
+    }
+
+    fn transition_to_completed(&mut self) -> Result<WinCryptoDtlsEvent, WinCryptoError> {
+        let mut srtp_parameters = SecPkgContext_SrtpParameters::default();
+        let Some(security_ctx) = self.security_ctx.as_ref() else {
+            return Err(WinCryptoError("Security context missing".to_string()));
+        };
+
+        unsafe {
+            QueryContextAttributesW(
+                security_ctx as *const _,
+                SECPKG_ATTR_STREAM_SIZES,
+                &mut self.encrypt_message_input_sizes as *mut _ as *mut std::ffi::c_void,
+            )
+            .map_err(|e| WinCryptoError(format!("SECPKG_ATTR_STREAM_SIZES: {:?}", e)))?;
+
+            QueryContextAttributesW(
+                security_ctx as *const _,
+                SECPKG_ATTR(SECPKG_ATTR_SRTP_PARAMETERS),
+                &mut srtp_parameters as *mut _ as *mut std::ffi::c_void,
+            )
+            .map_err(|e| {
+                WinCryptoError(format!("QueryContextAttributesW Keying Material: {:?}", e))
+            })?;
+        }
+
+        let srtp_profile_id = u16::from_be(srtp_parameters.ProtectionProfile);
+        let keying_material_info = SecPkgContext_KeyingMaterialInfo {
+            cbLabel: DTLS_KEY_LABEL.len() as u16,
+            pszLabel: windows::core::PSTR(DTLS_KEY_LABEL.as_ptr() as *mut u8),
+            cbKeyingMaterial: srtp_keying_material_len(srtp_profile_id)?,
+            cbContextValue: 0,
+            pbContextValue: std::ptr::null_mut(),
+        };
+        let mut keying_material = SecPkgContext_KeyingMaterial::default();
+
+        let srtp_keying_material = unsafe {
+            SetContextAttributesW(
+                security_ctx as *const _,
+                SECPKG_ATTR_KEYING_MATERIAL_INFO,
+                &keying_material_info as *const _ as *const std::ffi::c_void,
+                std::mem::size_of::<SecPkgContext_KeyingMaterialInfo>() as u32,
+            )
+            .map_err(|e| {
+                WinCryptoError(format!("SetContextAttributesA Keying Material: {:?}", e))
+            })?;
+
+            QueryContextAttributesExW(
+                security_ctx as *const _,
+                SECPKG_ATTR(SECPKG_ATTR_KEYING_MATERIAL),
+                &mut keying_material as *mut _ as *mut std::ffi::c_void,
+                std::mem::size_of::<SecPkgContext_KeyingMaterial>() as u32,
+            )
+            .map_err(|e| {
+                WinCryptoError(format!(
+                    "QueryContextAttributesExW Keying Material: {:?}",
+                    e
+                ))
+            })?;
+
+            std::slice::from_raw_parts(
+                keying_material.pbKeyingMaterial,
+                keying_material.cbKeyingMaterial as usize,
+            )
+            .to_vec()
+        };
+
+        unsafe {
```

Comment:
Generally I think every use of `unsafe` should have a `// SAFETY: xxx` comment that explains why it is safe. See for instance some libavcodec bindings we did at work:

https://github.com/lookback/libavcodec/blob/4b68e670ae56ad494d7658e434f3e7729183ba2e/src/decoder.rs#L232

In str0m I'm trying to do this for every `unwrap()` as well. It's sort of similar in that it tries to explain the invariant that is being upheld.

Reply:
Alright. I'll take a stab at it.